src/sys/mman.rs

```diff
@@ -425,19 +426,15 @@ pub unsafe fn mmap<F: AsFd>(
     flags: MapFlags,
     f: Option<F>,
     offset: off_t,
-) -> Result<*mut c_void> {
+) -> Result<NonNull<c_void>> {
     let ptr =
         addr.map_or(std::ptr::null_mut(), |a| usize::from(a) as *mut c_void);
 
     let fd = f.map(|f| f.as_fd().as_raw_fd()).unwrap_or(-1);
     let ret =
         libc::mmap(ptr, length.into(), prot.bits(), flags.bits(), fd, offset);
-
-    if ret == libc::MAP_FAILED {
-        Err(Errno::last())
-    } else {
-        Ok(ret)
-    }
+        
+    NonNull::new(ret).ok_or_else(Errno::last)
```

Comment:
[`NonNull::new()`](https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.new) will ONLY return a `None` if its argument is a null pointer, the error case of `mmap`, whose return value `libc::MAP_FAILED` ((void *) -1 = 0xffffffff) is not null:

```rust
use std::ptr::NonNull;

fn main() {
    let ptr = NonNull::new(libc::MAP_FAILED);
    println!("{:?}", ptr);
}
```

```
$ cargo r -q
Some(0xffffffffffffffff)
```

which means that the current impl will return `Ok(NonNull(0xffffffffff))` even when the underlying `mmap()` failed

Reply:
I've updated the implementation considering this.